app/bundles/AssetBundle/Controller/PublicController.php

```diff
@@ -1,92 +1,140 @@
 <?php
 
+declare(strict_types=1);
+
 namespace Mautic\AssetBundle\Controller;
 
-use Mautic\CoreBundle\Controller\FormController as CommonFormController;
+use Doctrine\ORM\EntityNotFoundException;
+use Doctrine\ORM\NonUniqueResultException;
+use Doctrine\ORM\ORMException;
+use Mautic\AssetBundle\Entity\Asset;
+use Mautic\AssetBundle\Model\AssetModel;
+use Mautic\CoreBundle\Controller\AbstractFormController;
 use Mautic\CoreBundle\Helper\CoreParametersHelper;
+use Symfony\Component\HttpFoundation\File\Exception\FileNotFoundException;
 use Symfony\Component\HttpFoundation\RedirectResponse;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 
-class PublicController extends CommonFormController
+class PublicController extends AbstractFormController
 {
     /**
-     * @param string $slug
+     * Handles public download of assets by slug.
+     *
+     * This method performs the initial validation of the slug, retrieves the
+     *  corresponding Asset entity, and then delegates further response logic
+     *  (such as redirects, streaming, or access denial) to `createAssetResponse()`.
+     *
+     * @throws ORMException
+     */
+    public function downloadAction(
+        Request $request,
+        CoreParametersHelper $parametersHelper,
+        AssetModel $model,
+        string $slug,
+    ): Response {
+        try {
+            $entity = $model->getRepository()->findByIdAndAlias($slug);
+        } catch (NonUniqueResultException|EntityNotFoundException|\InvalidArgumentException) {
+            return $this->notFound();
+        }
+
+        return $this->createAssetResponse($request, $parametersHelper, $model, $entity);
+    }
+
+    /**
+     * Determines and returns the appropriate response based on the given Asset entity.
+     *
+     * Logic:
+     * - If entity is missing → return 404
+     * - If access is not allowed → track and return 401
+     * - If remote asset → track and redirect to remote URL
+     * - If local asset → track and stream file
+     *
+     * @throws ORMException
+     */
+    private function createAssetResponse(Request $request, CoreParametersHelper $parametersHelper, AssetModel $model, ?Asset $entity): Response
+    {
+        if (!$this->isAccessAllowed($entity)) {
+            $model->trackDownload($entity, $request, 401);
+            $response = $this->accessDenied();
+        } elseif ($entity->isRemote()) {
+            $response = $this->remoteRedirectResponse($model, $entity, $request);
+        } else {
+            $response = $this->localDownloadResponse($model, $entity, $request, $parametersHelper);
+        }
+
+        return $response;
```

Comment:
Again, nitpick, but this reads better by following the "return early" rule:
```suggestion
        if (!$this->isAccessAllowed($entity)) {
            $model->trackDownload($entity, $request, 401);

            return $this->accessDenied();
        } elseif ($entity->isRemote()) {
            return $this->remoteRedirectResponse($model, $entity, $request);
        }

        return $this->localDownloadResponse($model, $entity, $request, $parametersHelper);
```